### PR TITLE
run.md: --expose does NOT override Dockerfile EXPOSE

### PR DESCRIPTION
**Pull request by SpencerBrown**
run.md states that the operator can override all defaults set in the Dockerfile, and explicitly says that `--expose` overrides the `EXPOSE` instruction. Neither of these are true. An `EXPOSE` instruction cannot be overridden, `--expose` can only add additional exposed ports.

This change fixes the instructions, and also takes the liberty of crisping up the grammar and phrasing in a place or two.

### CODE REVIEW

File: docs/reference/run.md [hunk 1]
```diff
@@ -44,10 +44,9 @@ settings affect:
  * runtime constraints on CPU and memory
  * privileges and LXC configuration
  
-An image developer may set defaults for these same settings when they create the
-image using the `docker build` command. Operators, however, can override all
-defaults set by the developer using the `run` options.  And, operators can also
-override nearly all the defaults set by the Docker runtime itself.
+An image developer may set defaults for these same settings when she creates the
```
Comment (cpuguy83) [inline docs/reference/run.md]: Note, developer making comments on grammar/sentance structure, but...

```
An image developer may set defaults for these same settings during `docker build`. 
Operaters can override nearly all the defaults set in the image.
```

File: docs/reference/run.md [hunk 2]
```diff
@@ -981,12 +980,11 @@ or two examples of how to pass more parameters to that ENTRYPOINT:
 ## EXPOSE (incoming ports)
 
 The Dockerfile doesn't give much control over networking, only providing
-the `EXPOSE` instruction to give a hint to the operator about what
-incoming ports might provide services. The following options work with
-or override the Dockerfile's exposed defaults:
+the `EXPOSE` instruction to specify incoming ports that provide services.
```
Comment (cpuguy83) [inline docs/reference/run.md]: I think the original version is better here.

Comment (SpencerBrown) [inline docs/reference/run.md]: The original version is incorrect. The `EXPOSE` instruction does not "give a hint to the operator", it explicitly specifies exposed ports and cannot be overridden. And, none of the "following options" "override the Dockerfile's exposed defaults".

Comment (cpuguy83) [inline docs/reference/run.md]: `EXPOSE` is purely for service discovery, it does nothing at all to the container itself.

Comment (SpencerBrown) [inline docs/reference/run.md]: Not sure what you mean by "does nothing at all to the container". `EXPOSE` sets exposed ports on the container, You can see this by doing a `docker run` on an image built with `EXPOSE`, you will see the named ports listed as exposed ports. It's certainly true that they are not exposed _to the host_ unless `-p` or `-P` is used on `docker run`, but that's not the point. The point is that the documentation is making factually incorrect statements, and the changes in this PR corrects them.

Comment (cpuguy83) [inline docs/reference/run.md]: @SpencerBrown That's my point. These are hints, they are not requirements.

**Comment by LK4D4** [comment]
@cpuguy83 @moxiegirl @SpencerBrown what's up here?

**Comment by SpencerBrown** [comment]
As a new guy to Docker, I read the doc and it misled me. I raised an issue and @thaJeztah suggested I do a PR. So I did. @cpuguy83 does not agree that the doc is misleading and incorrect. I'm not sure how to proceed here. I figured I would just leave this PR for others to judge, I pretty much said my piece. But, here's a summary of how I got here.

From my POV, as a new guy reading the docs, I take them pretty literally. If it says "the operator can override" I assume it means I can override an EXPOSE instruction. I actually depended on this when building one of my first Dockerfiles and was unhappy to discover I could not actually override an EXPOSE, even though the docs said I could.

I see @cpuguy83 's point about "hints", now that I understand more, but for someone reading the docs, you should not assume they know the internals of how the system works. You need to be very explicit and clear. Calling it a "hint" is way too subtle for system documentation.

**Comment by moxiegirl** [comment]
@SpencerBrown I agree we need to be more explicit.  @cpuguy83 gave it a shot but admits to not having his best writing hat on. He's a sharp guy tho so let me see if I can come up with something that sorts it for everyone --- explicitly stated for you and technically accurate for him.

File: docs/reference/run.md [hunk 3]
```diff
@@ -21,7 +21,7 @@ executes `docker run`, she starts a process with its own file system,
 its own networking, and its own isolated process tree.  The
 [*Image*](/terms/image/#image) which starts the process may define
 defaults related to the binary to run, the networking to expose, and
-more, but `docker run` gives final control to the operator who starts
+more, but `docker run` gives additional control to the operator who starts
```
Comment (moxiegirl) [inline docs/reference/run.md]: @SpencerBrown @cpuguy83 try this new intro --- WDYT?

--> https://gist.github.com/moxiegirl/3acf6eb35fbe777d1ffd

# Docker run reference

Docker runs processes in isolated containers. A container is a process which
runs on a host. The host may be local or remote. When an operator executes
`docker run`, the container process that runs is isolated in that it has its
own file system, its own networking, and its own isolated process tree separate
from the host. 

This page details how to use the `docker run` command to define the container's
 resources at runtime. 

## General form

The basic `docker run` command takes this form:

    $ docker run [OPTIONS] IMAGE[:TAG|@DIGEST] [COMMAND] [ARG...]
    
The `docker run` command must specify an [*IMAGE*](/terms/image/#image) to derive
the container from. An image developer can define image defaults related to

 * detached or foreground running
 * container identification
 * network settings
 * runtime constraints on CPU and memory
 * privileges and LXC configuration

With the `docker run [OPTIONS]` an operator can add to or override the image
defaults set by a developer. And, additionally, operators can override nearly
all the defaults set by the Docker runtime itself. The operator's ability to
override image and Docker runtime defaults is why
[*run*](/reference/commandline/cli/#run) has more options than any other
`docker` command.

>**Note**: Depending on your Docker system configuration, you may be required to
preface the `docker run` command with `sudo`. To avoid having to use `sudo` with
the `docker` command, your system administrator can create a Unix group called
`docker` and add users to it. For more information about this configuration,
refer to the Docker installation documentation for your operating system.

File: docs/reference/run.md [hunk 4]
```diff
@@ -981,12 +980,11 @@ or two examples of how to pass more parameters to that ENTRYPOINT:
 ## EXPOSE (incoming ports)
```
Comment (moxiegirl) [inline docs/reference/run.md]: @SpencerBrown I can see why this confused you. It said the same thing several different ways.  Let me know if this is easier to follow. @cpuguy83 please check I didn't bork the technical bits.

--->  https://gist.github.com/moxiegirl/ee4dbfb64f1c0191df51

The following `run` command options work with container networking:

    --expose=[]: Expose a port or a range of ports inside the container,
                These are additional to those exposed by the `EXPOSE` instruction
    -P=false   : Publish all exposed ports to the host interfaces
    -p=[]      : Publish a container᾿s port or a range of ports to the host 
                   format: ip:hostPort:containerPort | ip::containerPort | hostPort:containerPort | containerPort
                   Both hostPort and containerPort can be specified as a range of ports. 
                   When specifying ranges for both, the number of container ports in the range must match the number of host ports in the range. (e.g., `-p 1234-1236:1234-1236/tcp`)
                   (use 'docker port' to see the actual mapping)
    --link=""  : Add link to another container (<name or id>:alias or <name or id>)


With the exception of the `EXPOSE` directive, an image developer hasn't got much
control over networking. The `EXPOSE` instruction defines the initial incoming
ports that provide services. These ports are available to processes inside the
container. An operator can use the `--expose` option to add to the exposed ports.

To expose an container's internal port, an operator can start the container with
the `-P` or `-p` flag or start the client container with `--link`. The exposed
port is accessible on the host and the ports are available to any client
that can reach the host.

The `-P` option publishes all the ports to the host interfaces. Docker binds each
exposed port to a random port on the host. The range of ports are within an
*ephemeral port range* defined by `/proc/sys/net/ipv4/ip_local_port_range`. Use
the `-p` flag to explicitly map a single port or range of ports. 

The port number inside the container (where the service listens) does not need
to match the port number exposed on the outside of the container
(where clients connect). For example, inside the container you might have an HTTP
service listening on port 80 (and so you `EXPOSE 80` in the Dockerfile),
but outside the container the port might be 42800. To find the
mapping between the host ports and the exposed ports, use `docker port`.

If the operator uses `--link` when starting a new client container, then the
client container can access the exposed port via a private networking interface.
 Docker will set some environment variables in the client container to help
indicate which interface and port to use. For more information on linking, see
[the guide on linking container together](/userguide/dockerlinks/ )

**Comment by SpencerBrown** [comment]
first gist https://gist.github.com/moxiegirl/3acf6eb35fbe777d1ffd:  lgtm, nice

second gist https://gist.github.com/moxiegirl/ee4dbfb64f1c0191df51: this is like 5x better than the previous. And it jibes completely with my now-improved understanding of Docker port mapping.

however, it "exposes" (pun intended) the lack of consistent terminology of the port mapping components.

example:
"An operator can use the --expose option to add to the exposed ports."
followed by:
"To expose an container's internal port, an operator can start the container with
the -P or -p flag or start the client container with --link."

Which is it? --expose exposes a port, or -P or -p exposes an internal port?

Then we see this formulation:
"The port number inside the container (where the service listens)" 
which is another way of saying the same thing.

I would suggest sticking with "expose" and "publish" to describe the port mapping components.

I took a shot at FIFY, look at the comment to the gist.

**Comment by thaJeztah** [comment]
> Which is it? --expose exposes a port, or -P or -p exposes an internal port?

That caught my eye too. I think "publish" should be used (but I defer to @moxiegirl here, perhaps she had a special reason here).

**Comment by moxiegirl** [comment]
I agree @SpencerBrown @thaJeztah the distinction makes for an easier read and goes with the p in the flag.  Haven't heard from @cpuguy83 --- just want to make sure he is cool with it too.

**Comment by cpuguy83** [comment]
LGTM

**Comment by cpuguy83** [comment]
Let me rephrase, looks really great @moxiegirl 
Thanks @SpencerBrown

**Comment by thaJeztah** [comment]
Thanks, @cpuguy83! (< -- @SpencerBrown don't be put off by his comments, he's a really nice guy in person, just likes his comments to be to-the-point :smile:)

@SpencerBrown could you rebase, and update the PR with the suggestions @moxiegirl provided?

**Comment by thaJeztah** [comment]
ping @SpencerBrown (see my previous comment: https://github.com/docker/docker/pull/14625#issuecomment-124836050) thanks in advance!

**Comment by moxiegirl** [comment]
@thaJeztah happy to carry for @SpencerBrown if necessary.

**Comment by thaJeztah** [comment]
@moxiegirl that'd be great (no worries, @SpencerBrown, your sign-off stays intact :-)) thanks!

**Comment by SpencerBrown** [comment]
hey all... I've been on vacation... @cpuguy83, it's all good... @moxiegirl, yes please take my suggestions on your gist and do what you will with them, I'm fine with whatever.

**Comment by thaJeztah** [comment]
@SpencerBrown if you want to update this PR yourself, please do! :heart: Just let us know if you want to :+1:

**Comment by cpuguy83** [comment]
ping, any update on this one? @SpencerBrown Would you like someone to carry this?

**Comment by thaJeztah** [comment]
Actually I started on a carry, but was held up with other things, lol; https://github.com/docker/docker/compare/master...thaJeztah:carry-14625

**Comment by thaJeztah** [comment]
Updated PR here https://github.com/docker/docker/pull/15675, so I'm closing this one. Additional changes can be discussed on the new PR if needed.

Thanks @SpencerBrown for the initial work!
